subprojects/core/src/main/java/org/gradle/execution/taskgraph/DefaultTaskExecutionPlan.java

```diff
@@ -484,36 +482,98 @@ public TaskInfo getTaskToExecute() {
+                final AtomicBoolean restartIteration = new AtomicBoolean();
+                final Set<String> lockedProjects = Sets.newHashSet();
                 while (iterator.hasNext()) {
-                    TaskInfo taskInfo = iterator.next();
-                    allTasksComplete = allTasksComplete && taskInfo.isComplete();
-                    if (taskInfo.isReady() && taskInfo.allDependenciesComplete() && canRunWithWithCurrentlyExecutedTasks(taskInfo)) {
-                        nextMatching = taskInfo;
-                        iterator.remove();
-                        break;
+                    final TaskInfo taskInfo = iterator.next();
+                    Project project = taskInfo.getTask().getProject();
+                    final String projectPath = ((ProjectInternal) project).getIdentityPath().toString();
+                    if (taskInfo.isReady() && taskInfo.allDependenciesComplete()) {
+                        try {
+                            final ProjectLock projectLock = projectLockService.getProjectLock(projectPath);
+                            boolean couldNotLock = !projectLock.tryWithProjectLock(new Runnable() {
+                                @Override
+                                public void run() {
+                                    boolean canExecute = false;
+                                    try {
+                                        if (lockedProjects.contains(projectLock.getProjectPath())) {
+                                            // If the state of project locks has changed, we want to restart
+                                            // task selection so that we can be sure that the earliest task
+                                            // from this project in the execution queue is selected first.
+                                            restartIteration.set(true);
+                                        } else {
+                                            if (canRunWithWithCurrentlyExecutedTasks(taskInfo)) {
+                                                selected.set(true);
+                                                iterator.remove();
+                                                if (taskInfo.allDependenciesSuccessful()) {
+                                                    taskInfo.startExecution();
+                                                    recordTaskStarted(taskInfo);
+                                                    canExecute = true;
+                                                } else {
+                                                    taskInfo.skipExecution();
+                                                }
+                                            }
+                                        }
+                                    } finally {
+                                        // We need to unlock the plan before unlocking the project.  This prevents a deadlock
```

Comment:
I just meant that right now we have methods on `ProjectLock` that lock/unlock in a single call.  It seems like what we are talking about would require us to have "asynchronous" lock() and unlock() calls which means we would have to worry about catching locks that haven't been released, etc.  I'm not 100% sure that's true without digging into it, but that's my first impression.  I'll explore this a little and see how it shapes up.